hpx/util/checkpoint.hpp

```diff
@@ -195,8 +201,28 @@ namespace util {
         return ist;
     }
 
-    // Function object for save_checkpoint
+    // Function objects for save_checkpoint
     namespace detail {
+        // Properly handle non clients
+        template <typename T,
+            typename U = typename std::enable_if<
+                !hpx::traits::is_client<T>::value>::type>
+        void arch_data(hpx::serialization::output_archive& ar, T&& t)
+        {
+            ar << t;
+        }
+        // Properly handle Clients to compoents
+        template <typename Client, typename Server,
+            typename U =
+                typename std::enable_if<traits::is_client<Client>::value>::type>
+        void arch_data(hpx::serialization::output_archive& ar,
+            hpx::components::client_base<Client, Server>&& c)
+        {
+            // Use shared pointer to serialize server
+            hpx::future<std::shared_ptr<Server>> f_ptr =
+                hpx::get_ptr<Server>(c.get_id());
+            ar << f_ptr.get();
```

Comment:
You should directly serialize the future here, otherwise you'll easily deadlock.

Reply:
@sithhell How would I get a deadlock here?

Reply:
Checkpointing will block until the future is ready. This might be a problem, doesn't have to.
There's special support for futures though which you might want to support anyway.

Reply:
Could you elaborate what problems might arise from a local only serialization (besides making it more asynchronous)?

Reply:
Depends on where you run the code. It should not lead to a deadlock, probably, when you just checkpoint a given object hierarchy. It all depends on what get_ptr is doing and whether the code that depends on the checkpointing to succeed is responsible for allowing the call to `get_ptr` or `c.get_id()` (which is another blocking call) succeed.

Reply:
get_ptr will succeed for local (to its invocation) objects only, and check-pointing does not call `get_id()`, iirc.

Reply:
I am not concerned that those calls might fail. I am mostly concerned that a operation that might follow checkpointing will only make those operations to return the result. This was a major concern for sending parcels, as said, not sure the same applies here. This is just something that caught my attention.

Reply:
@sithhell I confused about what changes you would like to see me make. Would you like to see the `ar << f_ptr.get()` wrapped in a dataflow?

Reply:
@sithhell , @hkaiser and I have come up with a way which should address the issue. Let me push that solution for your review.

Reply:
@sithhell  do these changes satisfy your request?